### DISCUSSION TITLE
Initial Value Problem: Flows through a DiGraph from multiple sources to multiple sinks.

### DISCUSSION THREAD

**Discussion by FGFischer** [networkx/networkx#5260] [Q&A] [answered]
I have a `DiGraph` with multiple nodes (as sources) and multiple nodes (as sinks). The sources and sinks are connected with each other through a large meshed DiGraph (lots of cycles). The goal is to model the flow via one-directional transportroutes. The "conservation of mass" is important as well. 

What my code currently does is: 
Step 1:It transforms the `DiGraph` into an `undirected Graph ` and then i use the `find_cycle ` function to find all cycles of the `undirected Graph` .
After this is done the code checks each cycle in the list of cycle, if its rotation is clockwise or counter-clockwise. 
The rotation is important in my case, because that is how i determine the direction of the flow through `DiGraph` with positive flows only.

Step 2: Then i set initial transport conditions for a _single source-single sink_ or _multi source-single sink_ or `single source-multi sink` network. The massflows to the sinks and from the sources are given. 

Step 3:After i've set an initial transport condition i solve the flowproblem, e.g. with the [Hardy-Cross-Method ](https://en.wikipedia.org/wiki/Hardy_Cross_method) and my Code returns a new DiGraph with the proper massflows through and directions of edges. 

My Problem now is **Step 2** with _multiple sources and multiple sinks_ .If i cant reduce the DiGraph to one of the 3 Cases mentioned in _Step 2_ the algorithms don't work as intended.
I ve read the Chapter about [flows](https://networkx.org/documentation/stable/reference/algorithms/flow.html) in the NetworkX documentation but none of the available algorithms seems to be a solution for my problem.
Is there any algorithm i could implement?

**Comment by dschult** [comment, marked as answer]
Would the [simplex_method](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.flow.network_simplex.html#networkx.algorithms.flow.network_simplex) work? I think that could enable you to avoid steps 1 and 2.

Make the DiGraph, set the node demands (source +, sink -). Set the edge capacities if needed (you don't mention that so you can set them to some large value). Set the edge weights (the length or cost or other feature of the edges -- if irrelevant, set them all to 1).  The function finds the flow that minimizes the total cost of the flow while satisfying the demands and the capacities.

**Reply by FGFischer** [reply]
Thank you very much for your reply, 

I ve tried the network_simplex method and it worked very well. even for my bigger testing-network.

One little correction for future readers, about the [Network_Simplex](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.flow.network_simplex.html#networkx.algorithms.flow.network_simplex) method and especially the logic.
> Demand:
Nodes of the graph G are expected to have an attribute demand that indicates how much flow a node wants to **send (negative demand)** or **receive (positive demand)**. Note that the sum of the demands should be 0 otherwise the problem in not feasible. If this attribute is not present, a node is considered to have 0 demand. Default value: ‘demand’.

Thus, **sources are negative** and **sinks are positive**. I dont understand the logic behind this decision.

And for this very reason i got the following error message and couldn't figure out where my mistake was:
``networkx.exception.NetworkXUnfeasible: no flow satisfies all node demands``

**Reply by dschult** [reply]
Good point -- and its a tricky call.  Some fields come with the idea of "demand for the stuff that is flowing". Others come with the idea of sources and sinks. The numbers should change sign for the two cases. Luckily, the impact of having the wrong sign should only be t hat the flow is the same magnitude and opposite direction of what you wanted. So, 1) it is hopefully obviously backwards and 2) it is easy to flip the sign of either the flow solution, or to flip the sign of the demand and rerun.  But changing the sign  shouldn't affect the sum needing to be zero.
